From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bitbloxhub <45184892+bitbloxhub@users.noreply.github.com>
Date: Tue, 31 Jan 2023 16:03:58 -0800
Subject: [PATCH] Movable BlockEntities from carpet

wip dont use

diff --git a/src/main/java/chikacraft/ChikaWorldConfig.java b/src/main/java/chikacraft/ChikaWorldConfig.java
index 2274c544bae0790ed2ff3ecff931f757d289f75c..ce20a6dce540d161b264e201921a8112413b9285 100644
--- a/src/main/java/chikacraft/ChikaWorldConfig.java
+++ b/src/main/java/chikacraft/ChikaWorldConfig.java
@@ -90,7 +90,9 @@ public class ChikaWorldConfig {
     }
 
     public boolean mechanicRenewableSponges = false;
+    public boolean mechanicMovableBlockEntities = false;
     private void mechanicsSettings() {
         mechanicRenewableSponges = getBoolean("settings.mechanics.renewablesponges", mechanicRenewableSponges);
+        mechanicMovableBlockEntities = getBoolean("settings.mechanics.movableblockentities", mechanicMovableBlockEntities);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 20c2621a83f016af2be5380bf869936129f5f6bb..0e3ab8debfd879b6351397e4922b85406e724802 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -53,7 +53,9 @@ import net.minecraft.world.level.biome.BiomeManager;
 import net.minecraft.world.level.block.BaseFireBlock;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.EntityBlock;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.LidBlockEntity;
 import net.minecraft.world.level.block.entity.TickingBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.border.BorderChangeListener;
@@ -1694,4 +1696,61 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
         return getWorld().getEnvironment() == org.bukkit.World.Environment.THE_END;
     }
     // Purpur end
+
+    // Chika start - movable blockentities from carpet
+    public boolean setBlockStateWithBlockEntity(BlockPos blockPos_1, BlockState blockState_1, BlockEntity newBlockEntity, int int_1) {
+        if (isOutsideBuildHeight(blockPos_1) || !this.isClientSide && isDebug()) return false;
+        LevelChunk worldChunk_1 = this.getChunkAt(blockPos_1);
+        Block block_1 = blockState_1.getBlock();
+
+        BlockState blockState_2;
+        if (newBlockEntity != null && block_1 instanceof EntityBlock) {
+            blockState_2 = (worldChunk_1).setBlockStateWithBlockEntity(blockPos_1, blockState_1, newBlockEntity, (int_1 & 64) != 0);
+            if (newBlockEntity instanceof LidBlockEntity) {
+                scheduleTick(blockPos_1, block_1, 5);
+            }
+        } else {
+            blockState_2 = worldChunk_1.setBlockState(blockPos_1, blockState_1, (int_1 & 64) != 0);
+        }
+
+        if (blockState_2 == null) {
+            return false;
+        } else {
+            BlockState blockState_3 = this.getBlockState(blockPos_1);
+
+            if (blockState_3 != blockState_2 && (blockState_3.getLightBlock((BlockGetter) this, blockPos_1) != blockState_2.getLightBlock((BlockGetter) this, blockPos_1) || blockState_3.getLightEmission() != blockState_2.getLightEmission() || blockState_3.useShapeForLightOcclusion() || blockState_2.useShapeForLightOcclusion())) {
+                //ProfilerFiller profiler = getProfiler();
+                //profiler.push("queueCheckLight");
+                this.getChunkSource().getLightEngine().checkBlock(blockPos_1);
+                //profiler.pop();
+            }
+
+            if (blockState_3 == blockState_1) {
+                if (blockState_2 != blockState_3) {
+                    this.setBlocksDirty(blockPos_1, blockState_2, blockState_3);
+                }
+
+                if ((int_1 & 2) != 0 && (!this.isClientSide || (int_1 & 4) == 0) && (this.isClientSide || worldChunk_1.getFullStatus() != null && worldChunk_1.getFullStatus().isOrAfter(ChunkHolder.FullChunkStatus.TICKING))) {
+                    this.sendBlockUpdated(blockPos_1, blockState_2, blockState_1, int_1);
+                }
+
+                if (!this.isClientSide && (int_1 & 1) != 0) {
+                    this.updateNeighborsAt(blockPos_1, blockState_2.getBlock());
+                    if (blockState_1.hasAnalogOutputSignal()) {
+                        updateNeighbourForOutputSignal(blockPos_1, block_1);
+                    }
+                }
+
+                if ((int_1 & 16) == 0) {
+                    int int_2 = int_1 & -34;
+                    blockState_2.updateIndirectNeighbourShapes(this, blockPos_1, int_2); // prepare
+                    blockState_1.updateNeighbourShapes(this, blockPos_1, int_2); // updateNeighbours
+                    blockState_1.updateIndirectNeighbourShapes(this, blockPos_1, int_2); // prepare
+                }
+                this.onBlockStateChange(blockPos_1, blockState_2, blockState_3);
+            }
+            return true;
+        }
+    }
+    // Chika end
 }
diff --git a/src/main/java/net/minecraft/world/level/block/ChestBlock.java b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
index 330ff3bc5fd8625e37b79e1204eddbe88de62c03..f1fb3ac1d40af53e9d8e120aa10157695f7284f8 100644
--- a/src/main/java/net/minecraft/world/level/block/ChestBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ChestBlock.java
@@ -423,4 +423,35 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
         }
 
     }
+
+    // Chika start - movable blockentities from carpet
+    public boolean isSticky(BlockState state) {
+        return true; // i guess?
+    }
+
+    public boolean isStickyToNeighbor(Level level, BlockPos pos, BlockState state, BlockPos neighborPos, BlockState neighborState, Direction dir, Direction moveDir) {
+        if (!level.chikaConfig.mechanicMovableBlockEntities) {
+            return false;
+        }
+        
+        if (!neighborState.is((Block)(Object)this)) {
+            return false;
+        }
+
+        ChestType type = state.getValue(ChestBlock.TYPE);
+        ChestType neighborType = neighborState.getValue(ChestBlock.TYPE);
+
+        if (type == ChestType.SINGLE || neighborType == ChestType.SINGLE) {
+            return false;
+        }
+        if (type != neighborType.getOpposite()) {
+            return false;
+        }
+
+        Direction facing = state.getValue(ChestBlock.FACING);
+        Direction neighborFacing = state.getValue(ChestBlock.FACING);
+
+        return facing == neighborFacing;
+    }
+    // Chika end
 }
diff --git a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
index efc1bdcaf141a2c2152173439d707f50feaaf91a..8bb5e820edf6bd43c9cbae304f6717a8101ef7d4 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -40,7 +40,7 @@ public abstract class BlockEntity {
     private final BlockEntityType<?> type;
     @Nullable
     protected Level level;
-    protected final BlockPos worldPosition;
+    protected BlockPos worldPosition; // Chika - movable blockentities from carpet
     protected boolean remove;
     private BlockState blockState;
 
@@ -317,4 +317,9 @@ public abstract class BlockEntity {
         return this.persistentLore;
     }
     // Purpur end
+
+    // Chika - movable blockentities from carpet
+    public void setCMPos(BlockPos newPos) {
+        worldPosition = newPos;
+    }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index f456ad8a74464414f69b616a48ee9a2c1cee4d90..15eb08f7c977555774b0bf96a33f2201d92f6a84 100644
--- a/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -21,6 +21,7 @@ import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.CommandBlock;
 import net.minecraft.world.level.block.DirectionalBlock;
 import net.minecraft.world.level.block.Mirror;
 import net.minecraft.world.level.block.Rotation;
@@ -59,6 +60,7 @@ public class PistonBaseBlock extends DirectionalBlock {
     protected static final VoxelShape UP_AABB = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 12.0D, 16.0D);
     protected static final VoxelShape DOWN_AABB = Block.box(0.0D, 4.0D, 0.0D, 16.0D, 16.0D, 16.0D);
     private final boolean isSticky;
+    private ThreadLocal<List<BlockEntity>> list1_BlockEntities = new ThreadLocal<>(); //Unneccessary ThreadLocal if client and server use different PistonBlock instances // Chika - movable blockentities from carpet
 
     public PistonBaseBlock(boolean sticky, BlockBehaviour.Properties settings) {
         super(settings);
@@ -244,7 +246,7 @@ public class PistonBaseBlock extends DirectionalBlock {
             }
             // Paper end
             world.setBlock(pos, iblockdata1, 20);
-            world.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(pos, iblockdata1, (BlockState) this.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.from3DDataValue(data & 7)), enumdirection, false, true)); // Paper - diff on change
+            world.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(pos, iblockdata1, (BlockState) this.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.from3DDataValue(data & 7)), enumdirection, false, true) /* null / * i think * / */); // Paper - diff on change
             world.blockUpdated(pos, iblockdata1.getBlock());
             iblockdata1.updateNeighbourShapes(world, pos, 2);
             if (this.isSticky) {
@@ -297,6 +299,27 @@ public class PistonBaseBlock extends DirectionalBlock {
         return true;
     }
 
+    // Chika start - movable blockentities from carpet
+    private static boolean isPushableBlockEntity(Block block)  {
+        return block != Blocks.ENDER_CHEST && block != Blocks.ENCHANTING_TABLE &&
+                       block != Blocks.END_GATEWAY && block != Blocks.END_PORTAL && block != Blocks.MOVING_PISTON  &&
+                       block != Blocks.SPAWNER;
+    }
+
+    private static boolean ifHasBlockEntity(BlockState blockState, Level world) {
+        if (!blockState.hasBlockEntity()) {
+            return false;
+        } else {
+            return !(world.chikaConfig.mechanicMovableBlockEntities && isPushableBlockEntity(blockState.getBlock()));
+        }
+    }
+
+    private static PushReaction moveGrindstones(BlockState blockState, Level world) {
+        if (world.chikaConfig.mechanicMovableBlockEntities && blockState.getBlock() == Blocks.GRINDSTONE) return PushReaction.NORMAL;
+        return blockState.getPistonPushReaction();
+    }
+    // Chika end
+
     public static boolean isPushable(BlockState state, Level world, BlockPos pos, Direction direction, boolean canBreak, Direction pistonDir) {
         if (pos.getY() >= world.getMinBuildHeight() && pos.getY() <= world.getMaxBuildHeight() - 1 && world.getWorldBorder().isWithinBounds(pos)) {
             if (state.isAir()) {
@@ -323,6 +346,13 @@ public class PistonBaseBlock extends DirectionalBlock {
                     } else if ((Boolean) state.getValue(PistonBaseBlock.EXTENDED)) {
                         return false;
                     }
+                    // Chika start - movable blockentities from carpet
+                    Block block_1 = state.getBlock();
+
+                    if (world.chikaConfig.mechanicMovableBlockEntities && block_1 instanceof CommandBlock) {
+                        return true;
+                    }
+                    // Chika end
 
                     return !state.hasBlockEntity();
                 }
@@ -350,6 +380,32 @@ public class PistonBaseBlock extends DirectionalBlock {
             List<BlockPos> list = pistonextendschecker.getToPush();
             List<BlockState> list1 = Lists.newArrayList();
 
+            // Chika start - movable blockentities from carpet
+            if (world.chikaConfig.mechanicMovableBlockEntities) {
+                list1_BlockEntities.set(Lists.newArrayList());
+                System.out.println(list1.size());
+                System.out.println(list1_BlockEntities.get().size());
+                for (int i = 0; i < list.size(); ++i) {
+                    System.out.println("a");
+                    BlockPos blockpos = list.get(i);
+                    System.out.println("b");
+                    BlockEntity blockEntity = (list1.get(i).hasBlockEntity()) ? world.getBlockEntity(blockpos) : null;
+                    System.out.println("c");
+                    list1_BlockEntities.get().add(blockEntity);
+                    System.out.println("d");
+                    if (blockEntity != null) {
+                        System.out.println("e");
+                        //hopefully this call won't have any side effects in the future, such as dropping all the BlockEntity's items
+                        //we want to place this same(!) BlockEntity object into the world later when the movement stops again
+                        world.removeBlockEntity(blockpos);
+                        System.out.println("f");
+                        blockEntity.setChanged();
+                    }
+                    System.out.println("g");
+                }
+            }
+            // Chika end
+
             for (int i = 0; i < list.size(); ++i) {
                 BlockPos blockposition2 = (BlockPos) list.get(i);
                 BlockState iblockdata = world.getBlockState(blockposition2);
@@ -440,7 +496,13 @@ public class PistonBaseBlock extends DirectionalBlock {
                     iblockdata1 = world.getBlockState(oldPos);
                     map.replace(oldPos, iblockdata1);
                 }
-                world.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(blockposition3, iblockdata2, allowDesync ? list1.get(k) : iblockdata1, dir, retract, false));
+                // Chika start - movable blockentities from carpet
+                BlockEntity blockEntityPiston = MovingPistonBlock.newMovingBlockEntity(blockposition3, iblockdata2, list1.get(k), dir, !retract, false); //aaajmarker
+                if (world.chikaConfig.mechanicMovableBlockEntities)
+                    ((PistonMovingBlockEntity) blockEntityPiston).setCarriedBlockEntity(list1_BlockEntities.get().get(k));
+                world.setBlockEntity(blockEntityPiston);
+                // Chika end
+                //world.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(blockposition3, iblockdata2, allowDesync ? list1.get(k) : iblockdata1, dir, retract, false));
                 if (!allowDesync) {
                     world.setBlock(oldPos, Blocks.AIR.defaultBlockState(), 2 | 4 | 16 | 1024); // set air to prevent later physics updates from seeing this block
                 }
diff --git a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
index 221c5d080d55326e458c1182823d6b49224ef498..9cfe9859bc799b8c8062040f4217ebae91096388 100644
--- a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -17,6 +17,7 @@ import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.EntityBlock;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.state.BlockState;
@@ -44,6 +45,12 @@ public class PistonMovingBlockEntity extends BlockEntity {
     private long lastTicked;
     private int deathTicks;
 
+    // Chika start - movable blockentities from carpet 
+    private BlockEntity carriedBlockEntity;
+    private boolean renderCarriedBlockEntity = false;
+    private boolean renderSet = false;
+    // Chika end
+
     public PistonMovingBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.PISTON, pos, state);
     }
@@ -100,6 +107,34 @@ public class PistonMovingBlockEntity extends BlockEntity {
     private BlockState getCollisionRelatedBlockState() {
         return !this.isExtending() && this.isSourcePiston() && this.movedState.getBlock() instanceof PistonBaseBlock ? Blocks.PISTON_HEAD.defaultBlockState().setValue(PistonHeadBlock.SHORT, Boolean.valueOf(this.progress > 0.25F)).setValue(PistonHeadBlock.TYPE, this.movedState.is(Blocks.STICKY_PISTON) ? PistonType.STICKY : PistonType.DEFAULT).setValue(PistonHeadBlock.FACING, this.movedState.getValue(PistonBaseBlock.FACING)) : this.movedState;
     }
+    
+    // Chika start - movable blockentities from carpet
+    public BlockEntity getCarriedBlockEntity() {
+        return carriedBlockEntity;
+    }
+    
+    public void setCarriedBlockEntity(BlockEntity blockEntity) {
+        this.carriedBlockEntity = blockEntity;
+        if (this.carriedBlockEntity != null) {
+            carriedBlockEntity.setCMPos(worldPosition);
+            // this might be little dangerous since pos is final for a hashing reason?
+            if (level != null) carriedBlockEntity.setLevel(level);
+        }
+    }
+
+    public boolean isRenderModeSet() {
+        return renderSet;
+    }
+    
+    public boolean getRenderCarriedBlockEntity() {
+        return renderCarriedBlockEntity;
+    }
+    
+    public void setRenderCarriedBlockEntity(boolean b) {
+        renderCarriedBlockEntity = b;
+        renderSet = true;
+    }
+    // Chika end
 
     private static void moveCollidedEntities(Level world, BlockPos pos, float f, PistonMovingBlockEntity blockEntity) {
         Direction direction = blockEntity.getMovementDirection();
@@ -255,6 +290,25 @@ public class PistonMovingBlockEntity extends BlockEntity {
         return this.movedState;
     }
 
+    // Chika start - movable blockentities from carpet
+    private static boolean movableTEsetBlockState0(Level world, BlockPos blockPos_1, BlockState blockAState_2, int int_1, PistonMovingBlockEntity pistonBlockEntity) {
+        if (!world.chikaConfig.mechanicMovableBlockEntities)
+            return world.setBlock(blockPos_1, blockAState_2, int_1);
+        else
+            return world.setBlockStateWithBlockEntity(blockPos_1, blockAState_2, pistonBlockEntity.getCarriedBlockEntity(), int_1);
+    }
+
+    private boolean movableTEsetBlockState1(Level world, BlockPos blockPos_1, BlockState blockState_2, int int_1) {
+        if (!world.chikaConfig.mechanicMovableBlockEntities)
+            return world.setBlock(blockPos_1, blockState_2, int_1);
+        else {
+            boolean ret = world.setBlockStateWithBlockEntity(blockPos_1, blockState_2, this.carriedBlockEntity, int_1);
+            this.carriedBlockEntity = null; //this will cancel the finishHandleBroken
+            return ret;
+        }
+    }
+    // Chika end
+
     public void finalTick() {
         if (this.level != null && (this.progressO < 1.0F || this.level.isClientSide)) {
             this.progress = 1.0F;
@@ -269,11 +323,22 @@ public class PistonMovingBlockEntity extends BlockEntity {
                     blockState = Block.updateFromNeighbourShapes(this.movedState, this.level, this.worldPosition);
                 }
 
-                this.level.setBlock(this.worldPosition, blockState, 3);
+                movableTEsetBlockState1(this.level, this.worldPosition, blockState, 3); // Chika - movable blockentities from carpet
                 this.level.neighborChanged(this.worldPosition, blockState.getBlock(), this.worldPosition);
+
+                // Chika start - movable blockentities from carpet
+                if (level.chikaConfig.mechanicMovableBlockEntities && this.carriedBlockEntity != null && this.level.getBlockState(this.worldPosition).getBlock() == Blocks.AIR) {
+                    BlockState blockState_2;
+                    if (this.isSourcePiston)
+                        blockState_2 = Blocks.AIR.defaultBlockState();
+                    else
+                        blockState_2 = Block.updateFromNeighbourShapes(this.movedState, this.level, this.worldPosition);
+                    this.level.setBlockStateWithBlockEntity(this.worldPosition, blockState_2, this.carriedBlockEntity, 3);
+                    this.level.destroyBlock(this.worldPosition, false, null);
+                }
+                // Chika end
             }
         }
-
     }
 
     public static void tick(Level world, BlockPos pos, BlockState state, PistonMovingBlockEntity blockEntity) {
@@ -288,14 +353,14 @@ public class PistonMovingBlockEntity extends BlockEntity {
                 if (world.getBlockState(pos).is(Blocks.MOVING_PISTON)) {
                     BlockState blockState = Block.updateFromNeighbourShapes(blockEntity.movedState, world, pos);
                     if (blockState.isAir()) {
-                        world.setBlock(pos, blockEntity.movedState, io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPistonDuplication ? 84 : (84 | 2)); // Paper - force notify (flag 2), it's possible the set type by the piston block (which doesn't notify) set this block to air
+                        movableTEsetBlockState0(world, pos, blockEntity.movedState, io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPistonDuplication ? 84 : (84 | 2), blockEntity); // Paper - force notify (flag 2), it's possible the set type by the piston block (which doesn't notify) set this block to air // Chika - movable blockentities from carpet
                         Block.updateOrDestroy(blockEntity.movedState, blockState, world, pos, 3);
                     } else {
                         if (blockState.hasProperty(BlockStateProperties.WATERLOGGED) && blockState.getValue(BlockStateProperties.WATERLOGGED)) {
                             blockState = blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(false));
                         }
 
-                        world.setBlock(pos, blockState, 67);
+                        movableTEsetBlockState0(world, pos, blockState, 67, blockEntity); // Chika - movable blockentities from carpet
                         world.neighborChanged(pos, blockState.getBlock(), pos);
                     }
                 }
@@ -323,6 +388,15 @@ public class PistonMovingBlockEntity extends BlockEntity {
         this.progressO = this.progress;
         this.extending = nbt.getBoolean("extending");
         this.isSourcePiston = nbt.getBoolean("source");
+        // Chika start - movable blockentities from carpet
+        if (level != null && level.chikaConfig.mechanicMovableBlockEntities && nbt.contains("carriedTileEntityCM", 10)) {
+            if (this.movedState.getBlock() instanceof EntityBlock)
+                this.carriedBlockEntity = ((EntityBlock) (this.movedState.getBlock())).newBlockEntity(worldPosition, movedState);//   this.world);
+            if (carriedBlockEntity != null) //Can actually be null, as BlockPistonMoving.createNewTileEntity(...) returns null
+                this.carriedBlockEntity.load(nbt.getCompound("carriedTileEntityCM"));
+            setCarriedBlockEntity(carriedBlockEntity);
+        }
+        // Chika end
     }
 
     @Override
@@ -333,6 +407,11 @@ public class PistonMovingBlockEntity extends BlockEntity {
         nbt.putFloat("progress", this.progressO);
         nbt.putBoolean("extending", this.extending);
         nbt.putBoolean("source", this.isSourcePiston);
+        // Chika start - movable blockentities from carpet
+        if (level != null && level.chikaConfig.mechanicMovableBlockEntities && this.carriedBlockEntity != null) {
+            nbt.put("carriedTileEntityCM", this.carriedBlockEntity.saveWithoutMetadata());
+        }
+        // Chika end
     }
 
     public VoxelShape getCollisionShape(BlockGetter world, BlockPos pos) {
diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 79bf9c277fe98df176113de39360fb34ad917577..2a50cbf270e0d8de5d31117870c44c1f7f52fd5b 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -40,6 +40,7 @@ import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntityTicker;
 import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.block.entity.TickingBlockEntity;
+import net.minecraft.world.level.block.piston.MovingPistonBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.EuclideanGameEventListenerRegistry;
 import net.minecraft.world.level.gameevent.GameEventListener;
@@ -558,6 +559,103 @@ public class LevelChunk extends ChunkAccess {
          */  // Paper - remove try catch
     }
 
+    // Chika start - movable blockentities from carpet
+    private BlockEntity ifGetBlockEntity(LevelChunk worldChunk, BlockPos blockPos_1, LevelChunk.EntityCreationType worldChunk$CreationType_1) {
+        if (!this.level.chikaConfig.mechanicMovableBlockEntities) {
+            return this.getBlockEntity(blockPos_1, LevelChunk.EntityCreationType.CHECK);
+        }
+        else {
+            return this.level.getBlockEntity(blockPos_1);
+        }
+    }
+
+    public BlockState setBlockStateWithBlockEntity(BlockPos blockPos_1, BlockState newBlockState, BlockEntity newBlockEntity, boolean boolean_1) {
+        int x = blockPos_1.getX() & 15;
+        int y = blockPos_1.getY();
+        int z = blockPos_1.getZ() & 15;
+        LevelChunkSection chunkSection = this.getSection(this.getSectionIndex(y));
+        if (chunkSection.hasOnlyAir())
+        {
+            if (newBlockState.isAir())
+            {
+                return null;
+            }
+        }
+        
+        boolean boolean_2 = chunkSection.hasOnlyAir();
+        BlockState oldBlockState = chunkSection.setBlockState(x, y & 15, z, newBlockState);
+        if (oldBlockState == newBlockState)
+        {
+            return null;
+        }
+        else
+        {
+            Block newBlock = newBlockState.getBlock();
+            Block oldBlock = oldBlockState.getBlock();
+            ((Heightmap) this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING)).update(x, y, z, newBlockState);
+            ((Heightmap) this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES)).update(x, y, z, newBlockState);
+            ((Heightmap) this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR)).update(x, y, z, newBlockState);
+            ((Heightmap) this.heightmaps.get(Heightmap.Types.WORLD_SURFACE)).update(x, y, z, newBlockState);
+            boolean boolean_3 = chunkSection.hasOnlyAir();
+            if (boolean_2 != boolean_3)
+            {
+                this.level.getChunkSource().getLightEngine().updateSectionStatus(blockPos_1, boolean_3);
+            }
+            
+            if (!this.level.isClientSide)
+            {
+                if (!(oldBlock instanceof MovingPistonBlock))//this is a movableTE special case, if condition wasn't there it would remove the blockentity that was carried for some reason
+                    oldBlockState.onRemove(this.level, blockPos_1, newBlockState, boolean_1);//this kills it
+            }
+            else if (oldBlock != newBlock && oldBlock instanceof EntityBlock)
+            {
+                this.level.removeBlockEntity(blockPos_1);
+            }
+            
+            if (chunkSection.getBlockState(x, y & 15, z).getBlock() != newBlock)
+            {
+                return null;
+            }
+            else
+            {
+                BlockEntity oldBlockEntity = null;
+                if (oldBlockState.hasBlockEntity())
+                {
+                    oldBlockEntity = this.getBlockEntity(blockPos_1, LevelChunk.EntityCreationType.CHECK);
+                    if (oldBlockEntity != null)
+                    {
+                        oldBlockEntity.setBlockState(oldBlockState);
+                        updateBlockEntityTicker(oldBlockEntity);
+                    }
+                }
+
+                if (oldBlockState.hasBlockEntity())
+                {
+                    if (newBlockEntity == null)
+                    {
+                        newBlockEntity = ((EntityBlock) newBlock).newBlockEntity(blockPos_1, newBlockState);
+                    }
+                    if (newBlockEntity != oldBlockEntity && newBlockEntity != null)
+                    {
+                        newBlockEntity.clearRemoved();
+                        this.level.setBlockEntity(newBlockEntity);
+                        newBlockEntity.setBlockState(newBlockState);
+                        updateBlockEntityTicker(newBlockEntity);
+                    }
+                }
+
+                if (!this.level.isClientSide)
+                {
+                    newBlockState.onPlace(this.level, blockPos_1, oldBlockState, boolean_1); //This can call setblockstate! (e.g. hopper does)
+                }
+                
+                this.unsaved = true; // shouldSave
+                return oldBlockState;
+            }
+        }
+    }
+    // Chika end 
+
     // CraftBukkit start
     @Nullable
     @Override
@@ -612,7 +710,7 @@ public class LevelChunk extends ChunkAccess {
                     }
 
                     if (iblockdata.hasBlockEntity()) {
-                        BlockEntity tileentity = this.getBlockEntity(blockposition, LevelChunk.EntityCreationType.CHECK);
+                        BlockEntity tileentity = ifGetBlockEntity(this, blockposition, LevelChunk.EntityCreationType.CHECK); // Chika - movable blockentities from carpet
 
                         if (tileentity == null) {
                             tileentity = ((EntityBlock) block).newBlockEntity(blockposition, iblockdata);
